### ISSUE TITLE
Circular buffer overflow although messages are settled

### ISSUE THREAD

**Issue by Gadi-Manor** [amqp/rhea#166]
I have an app which does the following: 
Received messages from a specific queue and resend them to dynamic queues depending on the message data.
 I'll give all the information although I guess not all is relevant.
Sender:
`autosettle = false;
snd_settle_mode = 0`

Receiver:
`autoaccept = false;
rcv_settle_mode = 0`

Once the message is delivered to the right queue I call `delivery.accept()`. Also, `settled` events of the sender are raised.

But in the end, when I look at `session.outgoing.deliveries.size` I see it keep increasing until it reaches the limit. Maybe I got things a little wrong but shouldn't the size be reduced for every settled message?

**Comment by Gadi-Manor** [comment]
Understood. 

> you need to settle the delivery yourself

By doing `context.delivery.settled = true` ?

And another question. Now when messages are settled the circular buffer overflow issue is resolved but there is a new one. My sender is not `sendable` because at some point it runs out of credit. 
I know about the credit at the receiver side but what exactly is the sender's credit and how do I overcome this problem?

**Comment by grs** [comment]
Yes, ```context.delivery.settled = true``` will settle the delivery locally.

Your sender is granted credit by the pricess it is sending to. Is it possible the queue you are sending to is full?

**Comment by Gadi-Manor** [comment]
I don't think so because if restart my application it successfully sends more messages.

**Comment by grs** [comment]
Instead of ```context.delivery.settled = true``` try ```context.delivery.update(true)```. That may help with the sendability of the sender but not if it is actually a lack of credit from the broker. Not exactly sure what pending corresponds to, I'd guess either it means unacknowledged or can't send because receiver has not given credit.

**Comment by grs** [comment]
It is ActiveMQ you are using, right? You can run the broker with env var PN_TRACE_FRM=1 to get a protocol trace, which is very helpful in diagnosing issues. It is a little verbose so ideally get as simple a reproducer or as short a run as possible,

**Comment by Gadi-Manor** [comment]
Thanks! 
What I'm thinking is that I try to send too many messages very fast and I don't give time for the credit to increase again (I'm not sure when exactly the credit increases). What I did for now is to check if the credit is close to it's end (e.g. credit < 50) and if so I put my app on hold - Not sending anymore messages - for a while (a second) and then continue. By doing it I'm now managing to send a lot more messages without any issues.
